{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "# from sklearn.model_selection import train_test_split\n",
    "# from sklearn.metrics import r2_score\n",
    "# from xgboost import XGBRegressor\n",
    "\n",
    "# # Assurez-vous que votre DataFrame df est bien chargé\n",
    "# csv_file_path = '../data/silver.csv'\n",
    "# df = pd.read_csv(csv_file_path)\n",
    "\n",
    "# # Séparation des caractéristiques (features) et de la cible (target)\n",
    "# X = df.drop(columns=['salary_in_usd'])\n",
    "# y = df['salary_in_usd']\n",
    "\n",
    "# # Séparation des données en ensembles d'entraînement (60%), de validation (20%) et de test (20%)\n",
    "# X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)\n",
    "# X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)\n",
    "\n",
    "# # Encodage des variables catégorielles\n",
    "# X_train = pd.get_dummies(X_train)\n",
    "# X_val = pd.get_dummies(X_val)\n",
    "# X_test = pd.get_dummies(X_test)\n",
    "\n",
    "# # Assurez-vous que les ensembles de validation et de test ont les mêmes colonnes que l'ensemble d'entraînement\n",
    "# X_val = X_val.reindex(columns=X_train.columns, fill_value=0)\n",
    "# X_test = X_test.reindex(columns=X_train.columns, fill_value=0)\n",
    "\n",
    "# # Création et entraînement du modèle XGBoost\n",
    "# model = XGBRegressor()\n",
    "# model.fit(X_train, y_train, eval_set=[(X_val, y_val)], early_stopping_rounds=10, verbose=False)\n",
    "\n",
    "# # Prédiction sur les ensembles de validation et de test\n",
    "# y_val_pred = model.predict(X_val)\n",
    "# y_test_pred = model.predict(X_test)\n",
    "\n",
    "# # Évaluation du modèle\n",
    "# r2_val = r2_score(y_val, y_val_pred)\n",
    "# r2_test = r2_score(y_test, y_test_pred)\n",
    "\n",
    "# print(f\"Validation R2: {r2_val}\")\n",
    "# print(f\"Test R2: {r2_test}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 243 candidates, totalling 729 fits\n",
      "Best parameters: {'colsample_bytree': 0.7, 'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200, 'subsample': 0.9}\n",
      "Validation R2: 0.2951529237390842\n",
      "Test R2: 0.2505060369671258\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "\nAll the 5 fits failed.\nIt is very likely that your model is misconfigured.\nYou can try to debug the error by setting error_score='raise'.\n\nBelow are more details about the failures:\n--------------------------------------------------------------------------------\n5 fits failed with the following error:\nTraceback (most recent call last):\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 895, in _fit_and_score\n    estimator.fit(X_train, y_train, **fit_params)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 730, in inner_f\n    return func(**kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1055, in fit\n    train_dmatrix, evals = _wrap_evaluation_matrices(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/sklearn.py\", line 521, in _wrap_evaluation_matrices\n    train_dmatrix = create_dmatrix(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/sklearn.py\", line 958, in _create_dmatrix\n    return QuantileDMatrix(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 730, in inner_f\n    return func(**kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 1529, in __init__\n    self._init(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 1588, in _init\n    it.reraise()\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 576, in reraise\n    raise exc  # pylint: disable=raising-bad-type\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 557, in _handle_exception\n    return fn()\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 641, in <lambda>\n    return self._handle_exception(lambda: self.next(input_data), 0)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 1280, in next\n    input_data(**self.kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 730, in inner_f\n    return func(**kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 624, in input_data\n    new, cat_codes, feature_names, feature_types = _proxy_transform(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 1315, in _proxy_transform\n    arr, feature_names, feature_types = _transform_pandas_df(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 490, in _transform_pandas_df\n    _invalid_dataframe_dtype(data)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 308, in _invalid_dataframe_dtype\n    raise ValueError(msg)\nValueError: DataFrame.dtypes for data must be int, float, bool or category. When categorical type is supplied, The experimental DMatrix parameter`enable_categorical` must be set to `True`.  Invalid columns:experience_level: object, employment_type: object, job_title: object, employee_residence: object, company_location: object, company_size: object\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m/home/utilisateur/Documents/Projets/data_science_job_salaries/model/train_model.ipynb Cell 2\u001b[0m line \u001b[0;36m8\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/utilisateur/Documents/Projets/data_science_job_salaries/model/train_model.ipynb#W1sZmlsZQ%3D%3D?line=79'>80</a>\u001b[0m mlflow\u001b[39m.\u001b[39msklearn\u001b[39m.\u001b[39mlog_model(final_model, \u001b[39m\"\u001b[39m\u001b[39mxgboost_model\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/utilisateur/Documents/Projets/data_science_job_salaries/model/train_model.ipynb#W1sZmlsZQ%3D%3D?line=81'>82</a>\u001b[0m \u001b[39m# Validation croisée avec le modèle final\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/home/utilisateur/Documents/Projets/data_science_job_salaries/model/train_model.ipynb#W1sZmlsZQ%3D%3D?line=82'>83</a>\u001b[0m cross_val_r2 \u001b[39m=\u001b[39m cross_val_score(final_model, X, y, cv\u001b[39m=\u001b[39;49m\u001b[39m5\u001b[39;49m, scoring\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mr2\u001b[39;49m\u001b[39m'\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/utilisateur/Documents/Projets/data_science_job_salaries/model/train_model.ipynb#W1sZmlsZQ%3D%3D?line=83'>84</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCross-validated R2: \u001b[39m\u001b[39m{\u001b[39;00mcross_val_r2\u001b[39m.\u001b[39mmean()\u001b[39m}\u001b[39;00m\u001b[39m ± \u001b[39m\u001b[39m{\u001b[39;00mcross_val_r2\u001b[39m.\u001b[39mstd()\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/home/utilisateur/Documents/Projets/data_science_job_salaries/model/train_model.ipynb#W1sZmlsZQ%3D%3D?line=84'>85</a>\u001b[0m mlflow\u001b[39m.\u001b[39mlog_metric(\u001b[39m\"\u001b[39m\u001b[39mcross_val_r2_mean\u001b[39m\u001b[39m\"\u001b[39m, cross_val_r2\u001b[39m.\u001b[39mmean())\n",
      "File \u001b[0;32m~/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/sklearn/utils/_param_validation.py:213\u001b[0m, in \u001b[0;36mvalidate_params.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    207\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    208\u001b[0m     \u001b[39mwith\u001b[39;00m config_context(\n\u001b[1;32m    209\u001b[0m         skip_parameter_validation\u001b[39m=\u001b[39m(\n\u001b[1;32m    210\u001b[0m             prefer_skip_nested_validation \u001b[39mor\u001b[39;00m global_skip_validation\n\u001b[1;32m    211\u001b[0m         )\n\u001b[1;32m    212\u001b[0m     ):\n\u001b[0;32m--> 213\u001b[0m         \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    214\u001b[0m \u001b[39mexcept\u001b[39;00m InvalidParameterError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    215\u001b[0m     \u001b[39m# When the function is just a wrapper around an estimator, we allow\u001b[39;00m\n\u001b[1;32m    216\u001b[0m     \u001b[39m# the function to delegate validation to the estimator, but we replace\u001b[39;00m\n\u001b[1;32m    217\u001b[0m     \u001b[39m# the name of the estimator by the name of the function in the error\u001b[39;00m\n\u001b[1;32m    218\u001b[0m     \u001b[39m# message to avoid confusion.\u001b[39;00m\n\u001b[1;32m    219\u001b[0m     msg \u001b[39m=\u001b[39m re\u001b[39m.\u001b[39msub(\n\u001b[1;32m    220\u001b[0m         \u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mparameter of \u001b[39m\u001b[39m\\\u001b[39m\u001b[39mw+ must be\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    221\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mparameter of \u001b[39m\u001b[39m{\u001b[39;00mfunc\u001b[39m.\u001b[39m\u001b[39m__qualname__\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m must be\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    222\u001b[0m         \u001b[39mstr\u001b[39m(e),\n\u001b[1;32m    223\u001b[0m     )\n",
      "File \u001b[0;32m~/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/sklearn/model_selection/_validation.py:719\u001b[0m, in \u001b[0;36mcross_val_score\u001b[0;34m(estimator, X, y, groups, scoring, cv, n_jobs, verbose, fit_params, params, pre_dispatch, error_score)\u001b[0m\n\u001b[1;32m    716\u001b[0m \u001b[39m# To ensure multimetric format is not supported\u001b[39;00m\n\u001b[1;32m    717\u001b[0m scorer \u001b[39m=\u001b[39m check_scoring(estimator, scoring\u001b[39m=\u001b[39mscoring)\n\u001b[0;32m--> 719\u001b[0m cv_results \u001b[39m=\u001b[39m cross_validate(\n\u001b[1;32m    720\u001b[0m     estimator\u001b[39m=\u001b[39;49mestimator,\n\u001b[1;32m    721\u001b[0m     X\u001b[39m=\u001b[39;49mX,\n\u001b[1;32m    722\u001b[0m     y\u001b[39m=\u001b[39;49my,\n\u001b[1;32m    723\u001b[0m     groups\u001b[39m=\u001b[39;49mgroups,\n\u001b[1;32m    724\u001b[0m     scoring\u001b[39m=\u001b[39;49m{\u001b[39m\"\u001b[39;49m\u001b[39mscore\u001b[39;49m\u001b[39m\"\u001b[39;49m: scorer},\n\u001b[1;32m    725\u001b[0m     cv\u001b[39m=\u001b[39;49mcv,\n\u001b[1;32m    726\u001b[0m     n_jobs\u001b[39m=\u001b[39;49mn_jobs,\n\u001b[1;32m    727\u001b[0m     verbose\u001b[39m=\u001b[39;49mverbose,\n\u001b[1;32m    728\u001b[0m     fit_params\u001b[39m=\u001b[39;49mfit_params,\n\u001b[1;32m    729\u001b[0m     params\u001b[39m=\u001b[39;49mparams,\n\u001b[1;32m    730\u001b[0m     pre_dispatch\u001b[39m=\u001b[39;49mpre_dispatch,\n\u001b[1;32m    731\u001b[0m     error_score\u001b[39m=\u001b[39;49merror_score,\n\u001b[1;32m    732\u001b[0m )\n\u001b[1;32m    733\u001b[0m \u001b[39mreturn\u001b[39;00m cv_results[\u001b[39m\"\u001b[39m\u001b[39mtest_score\u001b[39m\u001b[39m\"\u001b[39m]\n",
      "File \u001b[0;32m~/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/sklearn/utils/_param_validation.py:213\u001b[0m, in \u001b[0;36mvalidate_params.<locals>.decorator.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    207\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m    208\u001b[0m     \u001b[39mwith\u001b[39;00m config_context(\n\u001b[1;32m    209\u001b[0m         skip_parameter_validation\u001b[39m=\u001b[39m(\n\u001b[1;32m    210\u001b[0m             prefer_skip_nested_validation \u001b[39mor\u001b[39;00m global_skip_validation\n\u001b[1;32m    211\u001b[0m         )\n\u001b[1;32m    212\u001b[0m     ):\n\u001b[0;32m--> 213\u001b[0m         \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    214\u001b[0m \u001b[39mexcept\u001b[39;00m InvalidParameterError \u001b[39mas\u001b[39;00m e:\n\u001b[1;32m    215\u001b[0m     \u001b[39m# When the function is just a wrapper around an estimator, we allow\u001b[39;00m\n\u001b[1;32m    216\u001b[0m     \u001b[39m# the function to delegate validation to the estimator, but we replace\u001b[39;00m\n\u001b[1;32m    217\u001b[0m     \u001b[39m# the name of the estimator by the name of the function in the error\u001b[39;00m\n\u001b[1;32m    218\u001b[0m     \u001b[39m# message to avoid confusion.\u001b[39;00m\n\u001b[1;32m    219\u001b[0m     msg \u001b[39m=\u001b[39m re\u001b[39m.\u001b[39msub(\n\u001b[1;32m    220\u001b[0m         \u001b[39mr\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mparameter of \u001b[39m\u001b[39m\\\u001b[39m\u001b[39mw+ must be\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    221\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mparameter of \u001b[39m\u001b[39m{\u001b[39;00mfunc\u001b[39m.\u001b[39m\u001b[39m__qualname__\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m must be\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    222\u001b[0m         \u001b[39mstr\u001b[39m(e),\n\u001b[1;32m    223\u001b[0m     )\n",
      "File \u001b[0;32m~/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/sklearn/model_selection/_validation.py:450\u001b[0m, in \u001b[0;36mcross_validate\u001b[0;34m(estimator, X, y, groups, scoring, cv, n_jobs, verbose, fit_params, params, pre_dispatch, return_train_score, return_estimator, return_indices, error_score)\u001b[0m\n\u001b[1;32m    429\u001b[0m parallel \u001b[39m=\u001b[39m Parallel(n_jobs\u001b[39m=\u001b[39mn_jobs, verbose\u001b[39m=\u001b[39mverbose, pre_dispatch\u001b[39m=\u001b[39mpre_dispatch)\n\u001b[1;32m    430\u001b[0m results \u001b[39m=\u001b[39m parallel(\n\u001b[1;32m    431\u001b[0m     delayed(_fit_and_score)(\n\u001b[1;32m    432\u001b[0m         clone(estimator),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    447\u001b[0m     \u001b[39mfor\u001b[39;00m train, test \u001b[39min\u001b[39;00m indices\n\u001b[1;32m    448\u001b[0m )\n\u001b[0;32m--> 450\u001b[0m _warn_or_raise_about_fit_failures(results, error_score)\n\u001b[1;32m    452\u001b[0m \u001b[39m# For callable scoring, the return type is only know after calling. If the\u001b[39;00m\n\u001b[1;32m    453\u001b[0m \u001b[39m# return type is a dictionary, the error scores can now be inserted with\u001b[39;00m\n\u001b[1;32m    454\u001b[0m \u001b[39m# the correct key.\u001b[39;00m\n\u001b[1;32m    455\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mcallable\u001b[39m(scoring):\n",
      "File \u001b[0;32m~/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/sklearn/model_selection/_validation.py:536\u001b[0m, in \u001b[0;36m_warn_or_raise_about_fit_failures\u001b[0;34m(results, error_score)\u001b[0m\n\u001b[1;32m    529\u001b[0m \u001b[39mif\u001b[39;00m num_failed_fits \u001b[39m==\u001b[39m num_fits:\n\u001b[1;32m    530\u001b[0m     all_fits_failed_message \u001b[39m=\u001b[39m (\n\u001b[1;32m    531\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39mAll the \u001b[39m\u001b[39m{\u001b[39;00mnum_fits\u001b[39m}\u001b[39;00m\u001b[39m fits failed.\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    532\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mIt is very likely that your model is misconfigured.\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    533\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mYou can try to debug the error by setting error_score=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mraise\u001b[39m\u001b[39m'\u001b[39m\u001b[39m.\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    534\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mBelow are more details about the failures:\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m{\u001b[39;00mfit_errors_summary\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    535\u001b[0m     )\n\u001b[0;32m--> 536\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(all_fits_failed_message)\n\u001b[1;32m    538\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    539\u001b[0m     some_fits_failed_message \u001b[39m=\u001b[39m (\n\u001b[1;32m    540\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m{\u001b[39;00mnum_failed_fits\u001b[39m}\u001b[39;00m\u001b[39m fits failed out of a total of \u001b[39m\u001b[39m{\u001b[39;00mnum_fits\u001b[39m}\u001b[39;00m\u001b[39m.\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    541\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mThe score on these train-test partitions for these parameters\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    545\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mBelow are more details about the failures:\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m{\u001b[39;00mfit_errors_summary\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m    546\u001b[0m     )\n",
      "\u001b[0;31mValueError\u001b[0m: \nAll the 5 fits failed.\nIt is very likely that your model is misconfigured.\nYou can try to debug the error by setting error_score='raise'.\n\nBelow are more details about the failures:\n--------------------------------------------------------------------------------\n5 fits failed with the following error:\nTraceback (most recent call last):\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/sklearn/model_selection/_validation.py\", line 895, in _fit_and_score\n    estimator.fit(X_train, y_train, **fit_params)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 730, in inner_f\n    return func(**kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/sklearn.py\", line 1055, in fit\n    train_dmatrix, evals = _wrap_evaluation_matrices(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/sklearn.py\", line 521, in _wrap_evaluation_matrices\n    train_dmatrix = create_dmatrix(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/sklearn.py\", line 958, in _create_dmatrix\n    return QuantileDMatrix(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 730, in inner_f\n    return func(**kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 1529, in __init__\n    self._init(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 1588, in _init\n    it.reraise()\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 576, in reraise\n    raise exc  # pylint: disable=raising-bad-type\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 557, in _handle_exception\n    return fn()\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 641, in <lambda>\n    return self._handle_exception(lambda: self.next(input_data), 0)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 1280, in next\n    input_data(**self.kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 730, in inner_f\n    return func(**kwargs)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/core.py\", line 624, in input_data\n    new, cat_codes, feature_names, feature_types = _proxy_transform(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 1315, in _proxy_transform\n    arr, feature_names, feature_types = _transform_pandas_df(\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 490, in _transform_pandas_df\n    _invalid_dataframe_dtype(data)\n  File \"/home/utilisateur/Documents/Projets/data_science_job_salaries/venv/lib/python3.10/site-packages/xgboost/data.py\", line 308, in _invalid_dataframe_dtype\n    raise ValueError(msg)\nValueError: DataFrame.dtypes for data must be int, float, bool or category. When categorical type is supplied, The experimental DMatrix parameter`enable_categorical` must be set to `True`.  Invalid columns:experience_level: object, employment_type: object, job_title: object, employee_residence: object, company_location: object, company_size: object\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score\n",
    "from sklearn.metrics import r2_score\n",
    "from xgboost import XGBRegressor\n",
    "import mlflow\n",
    "import mlflow.sklearn\n",
    "\n",
    "# Assurez-vous que votre DataFrame df est bien chargé\n",
    "csv_file_path = '../data/silver.csv'\n",
    "df = pd.read_csv(csv_file_path)\n",
    "\n",
    "# Séparation des caractéristiques (features) et de la cible (target)\n",
    "X = df.drop(columns=['salary_in_usd'])\n",
    "y = df['salary_in_usd']\n",
    "\n",
    "# Séparation des données en ensembles d'entraînement (60%), de validation (20%) et de test (20%)\n",
    "X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.4, random_state=42)\n",
    "X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)\n",
    "\n",
    "# Encodage des variables catégorielles\n",
    "X_train = pd.get_dummies(X_train)\n",
    "X_val = pd.get_dummies(X_val)\n",
    "X_test = pd.get_dummies(X_test)\n",
    "\n",
    "# Assurez-vous que les ensembles de validation et de test ont les mêmes colonnes que l'ensemble d'entraînement\n",
    "X_val = X_val.reindex(columns=X_train.columns, fill_value=0)\n",
    "X_test = X_test.reindex(columns=X_train.columns, fill_value=0)\n",
    "\n",
    "# Initialisation de MLflow\n",
    "mlflow_tracking_uri = \"http://127.0.0.1:5000\" \n",
    "experiment_name = \"xgboost_salary_prediction\"\n",
    "mlflow.set_tracking_uri(mlflow_tracking_uri)\n",
    "mlflow.set_experiment(experiment_name)\n",
    "\n",
    "# Enregistrement des paramètres et métriques avec MLflow\n",
    "with mlflow.start_run() as run:\n",
    "    # Définir les paramètres de recherche\n",
    "    param_grid = {\n",
    "        'n_estimators': [50, 100, 200],\n",
    "        'learning_rate': [0.01, 0.05, 0.1],\n",
    "        'max_depth': [3, 4, 5],\n",
    "        'subsample': [0.7, 0.8, 0.9],\n",
    "        'colsample_bytree': [0.7, 0.8, 0.9]\n",
    "    }\n",
    "\n",
    "    # Création du modèle XGBoost\n",
    "    xgb_model = XGBRegressor()\n",
    "\n",
    "    # Utiliser GridSearchCV pour trouver les meilleurs hyperparamètres\n",
    "    grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid, scoring='r2', cv=3, verbose=1, n_jobs=-1)\n",
    "    grid_search.fit(X_train, y_train, eval_set=[(X_val, y_val)], verbose=False)\n",
    "\n",
    "    # Meilleurs paramètres trouvés par GridSearchCV\n",
    "    best_params = grid_search.best_params_\n",
    "    print(f\"Best parameters: {best_params}\")\n",
    "\n",
    "    # Enregistrer les hyperparamètres dans MLflow\n",
    "    mlflow.log_params(best_params)\n",
    "\n",
    "    # Entraîner le modèle final avec les meilleurs paramètres\n",
    "    final_model = XGBRegressor(**best_params)\n",
    "    final_model.fit(X_train, y_train, eval_set=[(X_val, y_val)], verbose=False)\n",
    "\n",
    "    # Prédiction sur les ensembles de validation et de test\n",
    "    y_val_pred = final_model.predict(X_val)\n",
    "    y_test_pred = final_model.predict(X_test)\n",
    "\n",
    "    # Évaluation du modèle\n",
    "    r2_val = r2_score(y_val, y_val_pred)\n",
    "    r2_test = r2_score(y_test, y_test_pred)\n",
    "\n",
    "    # Enregistrer les métriques dans MLflow\n",
    "    mlflow.log_metric(\"r2_val\", r2_val)\n",
    "    mlflow.log_metric(\"r2_test\", r2_test)\n",
    "\n",
    "    print(f\"Validation R2: {r2_val}\")\n",
    "    print(f\"Test R2: {r2_test}\")\n",
    "\n",
    "    # Enregistrer le modèle dans MLflow\n",
    "    mlflow.sklearn.log_model(final_model, \"xgboost_model\")\n",
    "\n",
    "    # Validation croisée avec le modèle final\n",
    "    cross_val_r2 = cross_val_score(final_model, X, y, cv=5, scoring='r2')\n",
    "    print(f\"Cross-validated R2: {cross_val_r2.mean()} ± {cross_val_r2.std()}\")\n",
    "    mlflow.log_metric(\"cross_val_r2_mean\", cross_val_r2.mean())\n",
    "    mlflow.log_metric(\"cross_val_r2_std\", cross_val_r2.std())\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
